File: apps/cli/src/subagents/worktree.ts
/**
 * Git worktree 管理 — 为 subagent 创建隔离工作区,对齐 AGENTS.md 第 12 节
 * "多 Subagent 并行开发强制规则"的分支隔离要求(每个 Subagent 在 subagent/<id> 分支工作)。
 *
 * 做减法:
 *   - 仅封装 git worktree add / list / remove / prune,不维护本地缓存
 *   - 路径默认 .worktrees/<id>(相对 sourcePath),由调用方保证 .gitignore 已忽略
 *   - 遵守 AGENTS.md 第 8 节删除安全规则:removeWorktree 前必须确认 subagent 已结束
 *   - git 操作用 child_process spawnSync,不引入新依赖
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { spawnSync } from 'node:child_process';
import {
  createWorktree as createWorktreeCoW,
} from '../worktree.js';

export interface WorktreeInfo {
  path: string;
  branch: string;
  parentId: string;
  subagentId: string;
  createdAt: string;
}

const WORKTREE_DIR_ENV = 'IHUI_WORKTREE_DIR';

export function getDefaultWorktreeRoot(sourcePath: string): string {
  if (process.env[WORKTREE_DIR_ENV]) return process.env[WORKTREE_DIR_ENV]!;
  return path.join(sourcePath, '.worktrees');
}

export function worktreeBranchName(subagentId: string): string {
  return `subagent/${subagentId}`;
}

interface GitResult {
  ok: boolean;
  stdout: string;
  stderr: string;
}

function runGit(args: string[], cwd: string): GitResult {
  const r = spawnSync('git', args, {
    cwd,
    encoding: 'utf-8',
    windowsHide: true,
  });
  return {
    ok: r.status === 0,
    stdout: typeof r.stdout === 'string' ? r.stdout.trim() : '',
    stderr: typeof r.stderr === 'string' ? r.stderr.trim() : '',
  };
}

function ensureGitignoreEntry(sourcePath: string, entry: string): void {
  const gitignorePath = path.join(sourcePath, '.gitignore');
  let content = '';
  if (fs.existsSync(gitignorePath)) {
    content = fs.readFileSync(gitignorePath, 'utf-8');
  }
  if (content.split(/\r?\n/).includes(entry)) return;
  const newline = content.length > 0 && !content.endsWith('\n') ? '\n' : '';
  fs.writeFileSync(gitignorePath, content + newline + entry + '\n', 'utf-8');
}

export function createWorktree(
  parentId: string,
  subagentId: string,
  sourcePath: string,
): WorktreeInfo {
  const root = getDefaultWorktreeRoot(sourcePath);
  if (!fs.existsSync(root)) {
    fs.mkdirSync(root, { recursive: true });
  }
  ensureGitignoreEntry(sourcePath, '.worktrees/');

  const wtPath = path.join(root, subagentId);
  const branch = worktreeBranchName(subagentId);

  if (fs.existsSync(wtPath)) {
    throw new Error(`worktree 路径已存在: ${wtPath}`);
  }

  const head = runGit(['rev-parse', '--abbrev-ref', 'HEAD'], sourcePath);
  const startPoint = head.ok ? head.stdout : 'HEAD';

  const addResult = runGit(
    ['worktree', 'add', '-b', branch, wtPath, startPoint],
    sourcePath,
  );
  if (!addResult.ok) {
    throw new Error(`git worktree add 失败: ${addResult.stderr || addResult.stdout}`);
  }

  return {
    path: wtPath,
    branch,
    parentId,
    subagentId,
    createdAt: new Date().toISOString(),
  };
}

export function listWorktrees(sourcePath: string): WorktreeInfo[] {
  const r = runGit(['worktree', 'list', '--porcelain'], sourcePath);
  if (!r.ok) return [];
  const out: WorktreeInfo[] = [];
  let curPath = '';
  let curBranch = '';
  for (const line of r.stdout.split(/\r?\n/)) {
    if (line.startsWith('worktree ')) {
      curPath = line.slice('worktree '.length).trim();
    } else if (line.startsWith('branch ')) {
      curBranch = line.slice('branch '.length).trim();
    } else if (line === '' && curPath && curBranch.startsWith('subagent/')) {
      const subagentId = curBranch.slice('subagent/'.length);
      out.push({
        path: curPath,
        branch: curBranch,
        parentId: '',
        subagentId,
        createdAt: '',
      });
      curPath = '';
      curBranch = '';
    }
  }
  if (curPath && curBranch.startsWith('subagent/')) {
    const subagentId = curBranch.slice('subagent/'.length);
    out.push({
      path: curPath,
      branch: curBranch,
      parentId: '',
      subagentId,
      createdAt: '',
    });
  }
  return out;
}

export interface RemoveWorktreeOptions {
  sourcePath: string;
  force?: boolean;
}

export function removeWorktree(
  wtPath: string,
  opts: RemoveWorktreeOptions,
): boolean {
  if (!fs.existsSync(wtPath)) return false;
  const args = ['worktree', 'remove', wtPath];
  if (opts.force) args.push('--force');
  const r = runGit(args, opts.sourcePath);
  if (!r.ok) {
    const rm = spawnSync('rm', ['-rf', wtPath], { windowsHide: true });
    if (rm.error || rm.status !== 0) {
      throw new Error(`worktree 删除失败: ${r.stderr || r.stdout}`);
    }
  }
  return true;
}

export function pruneWorktrees(sourcePath: string): number {
  const before = listWorktrees(sourcePath);
  const r = runGit(['worktree', 'prune'], sourcePath);
  if (!r.ok) return 0;
  const after = listWorktrees(sourcePath);
  return Math.max(0, before.length - after.length);
}

/**
 * 创建 worktree(快路径优先,失败 fallback 到原 git worktree add)。
 *
 * 策略(P1-8 Fast Worktree CoW 集成):
 *   - useFastPath=false(默认):直接调原 createWorktree(git worktree add -b),零回归
 *   - useFastPath=true:先尝试 CoW 复制 + git checkout -b 建分支,失败则清理半成品目录后
 *     fallback 到原 createWorktree,保证最终一致
 *
 * CoW 路径与原路径语义对齐:都生成 subagent/<id> 分支、相同的 destination 路径、
 * 相同的 .gitignore 注入。区别仅在于 .git 是独立拷贝(非 linked worktree),
 * 因此不会被 git worktree list 列出,但 removeWorktree 会 fallback 到 rm -rf 清理。
 *
 * @param parentId 父会话 ID
 * @param subagentId subagent ID
 * @param sourcePath 源仓库路径
 * @param useFastPath 是否启用 CoW 快路径(feature flag,默认 false)
 * @returns WorktreeInfo + usedFastPath 标记 + cowKind(仅快路径成功时有值)
 */
export async function createWorktreeWithFallback(
  parentId: string,
  subagentId: string,
  sourcePath: string,
  useFastPath: boolean = false,
): Promise<WorktreeInfo & { usedFastPath: boolean; cowKind?: string }> {
  if (useFastPath) {
    const root = getDefaultWorktreeRoot(sourcePath);
    if (!fs.existsSync(root)) {
      fs.mkdirSync(root, { recursive: true });
    }
    ensureGitignoreEntry(sourcePath, '.worktrees/');
    const destination = path.join(root, subagentId);
    const branch = worktreeBranchName(subagentId);

    // 路径已存在时跳过 CoW(交给原函数抛出统一的“路径已存在”错误)
    if (!fs.existsSync(destination)) {
      try {
        // CoW 复制源工作树(含 .git),preserveWorkingTree=false 会 reset --hard
        const result = await createWorktreeCoW({
          source: sourcePath,
          destination,
          preserveWorkingTree: false,
        });

        // CoW 复制后 HEAD 与源一致,需创建 subagent/<id> 分支(等价 git worktree add -b)
        const checkout = runGit(['checkout', '-b', branch], destination);
        if (!checkout.ok) {
          throw new Error(
            `git checkout -b ${branch} 失败: ${checkout.stderr || checkout.stdout}`,
          );
        }

        return {
          path: result.destination,
          branch,
          parentId,
          subagentId,
          createdAt: new Date().toISOString(),
          usedFastPath: true,
          cowKind: result.cowKind,
        };
      } catch {
        // CoW 失败 → 清理半成品目录后 fallback 到原 git worktree add
        try {
          if (fs.existsSync(destination)) {
            fs.rmSync(destination, { recursive: true, force: true });
          }
        } catch {
          // 清理失败忽略,原 git worktree add 会因路径存在抛错
        }
      }
    }
  }

  // 原有逻辑:git worktree add -b <branch> <path> <startPoint>
  const wt = createWorktree(parentId, subagentId, sourcePath);
  return { ...wt, usedFastPath: false };
}

File: apps/ai-service/app/services/worktree.py
"""Git worktree 隔离(P1-2,对齐 CLI 端 apps/cli/src/subagents/worktree.ts)。

用 asyncio.create_subprocess_exec 调 git(零新依赖,跨平台)。
每个 executor 在独立 worktree 中跑,防多 worker 并发改同一文件冲突。

Windows 注意:
- 路径长度:worktree 创建后 git config core.longpaths true
- 文件锁:删除失败 fallback shutil.rmtree(ignore_errors=True)
- 符号链接:git config core.symlinks false(Windows 不稳定)
- 大小写:task_id 统一小写(ULID/UUID .lower())
"""

import asyncio
import logging
import os
import re
import shutil
from dataclasses import dataclass
from typing import Optional

logger = logging.getLogger(__name__)

# worktree 根目录环境变量(对齐 CLI 端 WORKTREE_DIR_ENV)
WORKTREE_DIR_ENV = "IHUI_WORKTREE_DIR"
# task_id 路径安全校验(防 ../ 注入)
_TASK_ID_RE = re.compile(r"^[A-Za-z0-9_-]+$")


@dataclass
class WorktreeInfo:
    """worktree 信息(对齐 CLI 端 WorktreeInfo)。"""

    path: str
    branch: str
    parent_id: str  # 源仓库路径


def get_default_worktree_root(source_path: str) -> str:
    """获取 worktree 根目录(对齐 CLI 端 getWorktreeRoot)。"""
    env = os.environ.get(WORKTREE_DIR_ENV)
    return env if env else os.path.join(source_path, ".worktrees")


def worktree_path(source_path: str, task_id: str) -> str:
    """计算 worktree 路径(.worktrees/<task_id>)。"""
    safe_id = task_id.lower() if not _TASK_ID_RE.match(task_id) else task_id
    return os.path.join(get_default_worktree_root(source_path), safe_id)


async def _git(
    args: list[str], cwd: str, *, timeout: float = 30.0
) -> tuple[int, bytes, bytes]:
    """调 git 子进程,返回 (returncode, stdout, stderr)。"""
    proc = await asyncio.create_subprocess_exec(
        "git",
        *args,
        cwd=cwd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    try:
        stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout)
    except asyncio.TimeoutError:
        proc.kill()
        await proc.wait()
        raise
    return proc.returncode or 0, stdout, stderr


async def ensure_gitignore(source_path: str) -> None:
    """确保 .worktrees/ 在 .gitignore 中(对齐 CLI 端 ensureGitignore)。"""
    gitignore_path = os.path.join(source_path, ".gitignore")
    try:
        with open(gitignore_path, "r", encoding="utf-8") as f:
            content = f.read()
        if ".worktrees/" not in content:
            with open(gitignore_path, "a", encoding="utf-8") as f:
                if not content.endswith("\n"):
                    f.write("\n")
                f.write(".worktrees/\n")
    except FileNotFoundError:
        with open(gitignore_path, "w", encoding="utf-8") as f:
            f.write(".worktrees/\n")


async def create_worktree(
    source_path: str, task_id: str, *, start_point: Optional[str] = None
) -> WorktreeInfo:
    """创建 worktree(对齐 CLI 端 createWorktree)。

    Args:
        source_path: 源仓库路径
        task_id: 任务 ID(用作 worktree 目录名 + 分支名)
        start_point: 起始点(默认当前 HEAD)

    Returns:
        WorktreeInfo(path, branch, parent_id)

    Raises:
        RuntimeError: git worktree add 失败
    """
    if not _TASK_ID_RE.match(task_id):
        raise ValueError(f"task_id 含非法字符: {task_id}")

    await ensure_gitignore(source_path)
    wt_path = worktree_path(source_path, task_id)
    branch = f"subagent/{task_id}"

    # 探测 startPoint(对齐 CLI 端:先 rev-parse --abbrev-ref HEAD,失败用 HEAD)
    if start_point is None:
        rc, stdout, _ = await _git(
            ["rev-parse", "--abbrev-ref", "HEAD"], cwd=source_path
        )
        start_point = stdout.decode().strip() if rc == 0 and stdout.strip() else "HEAD"

    # git worktree add -b <branch> <path> <startPoint>
    rc, _, stderr = await _git(
        ["worktree", "add", "-b", branch, wt_path, start_point],
        cwd=source_path,
        timeout=60.0,
    )
    if rc != 0:
        raise RuntimeError(
            f"git worktree add 失败(branch={branch}, path={wt_path}): {stderr.decode().strip()}"
        )

    # Windows 适配:长路径 + 关闭 symlink(对齐研究建议)
    if os.name == "nt":
        await _git(["config", "core.longpaths", "true"], cwd=wt_path, timeout=5.0)
        await _git(["config", "core.symlinks", "false"], cwd=wt_path, timeout=5.0)

    return WorktreeInfo(path=wt_path, branch=branch, parent_id=source_path)


async def remove_worktree(
    wt_path: str, source_path: str, *, force: bool = True
) -> None:
    """删除 worktree(对齐 CLI 端 removeWorktree)。

    三层 fallback:
    1. git worktree remove --force(从源仓库 cwd 调,清理元数据)
    2. shutil.rmtree(ignore_errors=True)(git 失败时,如文件锁)
    3. 仅 log warning(全失败时不抛错,避免阻塞 WorkerPool)
    """
    # 层 1:git worktree remove
    args = ["worktree", "remove"]
    if force:
        args.append("--force")
    args.append(wt_path)
    rc, _, stderr = await _git(args, cwd=source_path, timeout=30.0)
    if rc == 0:
        return

    # 层 2:shutil.rmtree(git 失败,如 Windows 文件锁)
    logger.warning(
        "git worktree remove 失败,fallback shutil.rmtree: %s",
        stderr.decode().strip() if stderr else "unknown",
    )
    shutil.rmtree(wt_path, ignore_errors=True)

    # 层 3:检查是否仍存在(可能文件锁未释放)
    if os.path.exists(wt_path):
        logger.warning(
            "worktree 仍存在(可能文件锁): %s,将由后续 prune 清理", wt_path
        )


async def prune_worktrees(source_path: str) -> None:
    """清理孤儿 worktree 元数据(对齐 CLI 端 pruneWorktrees)。

    WorkerPool start() 时调一次,清理上次进程崩溃残留。
    """
    await _git(["worktree", "prune"], cwd=source_path, timeout=30.0)


async def list_worktrees(source_path: str) -> list[WorktreeInfo]:
    """列出所有 worktree(对齐 CLI 端 listWorktrees)。"""
    rc, stdout, _ = await _git(
        ["worktree", "list", "--porcelain"], cwd=source_path, timeout=10.0
    )
    if rc != 0:
        return []
    result: list[WorktreeInfo] = []
    current_path: Optional[str] = None
    for line in stdout.decode().splitlines():
        if line.startswith("worktree "):
            # 前一个 worktree 可能没有 branch 行(detached HEAD),也要加入
            if current_path:
                result.append(
                    WorktreeInfo(path=os.path.normpath(current_path), branch="", parent_id=source_path)
                )
            current_path = line[len("worktree ") :]
        elif line.startswith("branch ") and current_path:
            branch_ref = line[len("branch ") :]
            # branch refs/heads/<name> → <name>
            branch_name = (
                branch_ref[len("refs/heads/") :]
                if branch_ref.startswith("refs/heads/")
                else branch_ref
            )
            result.append(
                WorktreeInfo(path=os.path.normpath(current_path), branch=branch_name, parent_id=source_path)
            )
            current_path = None
        elif line == "" and current_path:
            # 空行分隔,前一个 worktree 无 branch 行
            result.append(
                WorktreeInfo(path=os.path.normpath(current_path), branch="", parent_id=source_path)
            )
            current_path = None
    # 最后一个 worktree
    if current_path:
        result.append(
            WorktreeInfo(path=os.path.normpath(current_path), branch="", parent_id=source_path)
        )
    return result
